### PR TITLE
Reorganize the header files [1.1.1]

### PR DESCRIPTION
**Pull request by mspncp**
This pull request is the backport of #9333 to the OpenSSL_1_1_1-stable branch. All commits are autogenerated using the `reorganize-header-files` script from #9333 as follows:

```
./reorganize-header-files 111
```

### CODE REVIEW

**Comment by mspncp** [comment]
Before you start complaining: I see the new include guard names might need some more finetuning... ;-)

**Comment by mspncp** [comment]
Up-to-date with #9333 again.

**Comment by mspncp** [comment]
Synced with #9333, see [force-push diff](https://github.com/openssl/openssl/compare/dfc50b309e532c664856490d084f2c18102e9669..4889a09003ae62ce1967499086a291af5360dade).

**Comment by petrovr** [comment]
Hmm, rename of public headers in stable (release) branch!
Hell is full of good meanings.

**Comment by mspncp** [comment]
It is true that <openssl/ossl_typ.h> was renamed to <openssl/types.h>, but a substitute for the old header file was added which forwards to the new one.

**Comment by mspncp** [comment]
(I did the renaming and the replacement in two separate  commits in order not to confuse git‘s automatic move detection)

**Comment by mspncp** [comment]
Synchronized with #9333, no rebase.

**Comment by mspncp** [comment]
Rebased without changes.

File: include/openssl/aes.h [hunk 1]
```diff
@@ -7,8 +7,11 @@
  * https://www.openssl.org/source/license.html
  */
 
-#ifndef HEADER_AES_H
-# define HEADER_AES_H
+#ifndef OPENSSL_AES_H
+# define OPENSSL_AES_H
+# pragma once
+
+# define HEADER_AES_H /* deprecated in version 3.0 */
```
Comment (levitte) [inline include/openssl/aes.h]: Why do we talk about 3.0 deprecation this in 1.1.1?

File: include/openssl/ossl_typ.h [hunk 2]
```diff
@@ -1,196 +1,16 @@
 /*
- * Copyright 2001-2018 The OpenSSL Project Authors. All Rights Reserved.
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
  *
- * Licensed under the OpenSSL license (the "License").  You may not use
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
```
Comment (levitte) [inline include/openssl/ossl_typ.h]: No, do not change the license in 1.1.1

**Comment by levitte** [comment]
We're releasing 1.1.1d tomorrow.  Whatever else we decide, this PR should be merged *after*.

**Comment by mspncp** [comment]
> We're releasing 1.1.1d tomorrow. Whatever else we decide, this PR should be merged _after_.

Oh, certainly. I wouldn't this change either.

File: include/openssl/aes.h (hunk 1 above)
Comment (mspncp) [reply include/openssl/aes.h]: I'll remove the comment.

File: include/openssl/ossl_typ.h (hunk 2 above)
Comment (mspncp) [reply include/openssl/ossl_typ.h]: That change was not intentional of course.

**Comment by mspncp** [comment]
> Quite frankly, I'm apprehensive of changing the public headers at all in 1.1.1. The internal / private headers, go right ahead... but the public ones? Not too sure about those.

The primary goal of this backport is to facilitate future backports from master to 1.1.1. Since only the internal and private headers were heavily restructured, only those changes really need to be backported.

Omitting the changes to the public headers from the backport sounds like a reasonable compromise and is more compatible with the stability guarantees of our release strategy.

**Comment by levitte** [comment]
> Omitting the changes to the public headers from the backport sounds like a reasonable compromise and is more compatible with the stability guarantees of our release strategy.

Thanks, that's exactly what I was thinking.

**Comment by levitte** [comment]
Rebase (or re-run on fresher `OpenSSL_1_1_1-stable`)

**Comment by mspncp** [comment]
> Rebase (or re-run on fresher `OpenSSL_1_1_1-stable`)

Done.

**Comment by mspncp** [comment]
> Hmm, rename of public headers in stable (release) branch!
> Hell is full of good meanings.

@petrovr I hope you are enjoyed to hear that your admonishing words [changed my mind](https://github.com/openssl/openssl/pull/9681#issuecomment-529602249).

_(Oh, and BTW: [If There's a Hell Below We're All Going to Go](https://www.youtube.com/watch?v=x1xmXOP3lhM)...)_

**Comment by mspncp** [comment]
The previous force-push was a rebase without changing the script. The following force-push will be a rerun with the changed script, without rebasing.

In other words: the second force-push diff will contain all relevant changes. The most interesting modification is that the changes to the public headers are applied only to master but not to the 1.1.1 stable branch.

**Comment by mspncp** [comment]
No diffs in `include/openssl` anymore :-)

```
git diff OpenSSL_1_1_1-stable pr-reorganize-header-files-111 -- include/openssl
```

**Comment by mspncp** [comment]
> Quite frankly, I'm apprehensive of changing the public headers at all in 1.1.1. The internal / private headers, go right ahead... but the public ones? Not too sure about those.
> 
> @openssl/omc, @openssl/committers, I'd like a second opinion

All changes in include/openssl are not backported anymore. This can be checked using

```
git diff OpenSSL_1_1_1-stable pr-reorganize-header-files-111 -- include/openssl
```

**Comment by mspncp** [comment]
The unchanged script was rerun after rebasing on OpenSSL_1_1_1-stable

**Comment by mspncp** [comment]
Done.

**OpenSSL_1_1_1-stable**

ea21f6f88e Add util/fix-includes script
fbbfd128c9 Fix header file include guard names
b5acbf9148 Reorganize local header files
0c994d54af Reorganize private crypto header files

**Comment by richsalz** [comment]
Great job.  Congrats.

**Pull request closed without merging**